R/cube-transforms.R

```diff
@@ -0,0 +1,177 @@
+#' #               All opinions             Strongly Agree             Somewhat Agree 
+#' #                         NA                         23                         24 
+#' # Neither Agree nor Disagree          Strongly Disagree          Somewhat Disagree 
+#' #                         18                         19                         16 
+#' }
+#'
+#' @aliases subtotalArray
+#' @export
+applyTransforms <- function (x, ary = cubeToArray(x), ...) {
+    # if there are row transforms, calculate them and display them
+    row_trans <- tryCatch(Transforms(data=index(variables(x))[[1]]$view$transform), error = function(e) NULL)
+    if (!is.null(row_trans)) {
+        var_cats <- Categories(data=index(variables(x))[[1]]$categories)
+
+        # TODO: calculate category/element changes
+
+        if (length(dim(ary)) > 1) {
+            off_margins <- seq_along(dim(ary))[-1]
+            dim_names  <- names(dimnames(ary))
+            ary <- apply(ary, off_margins, calcTransforms, row_trans, var_cats, ...)
+            names(dimnames(ary)) <- dim_names
+        } else {
+            ary <- as.array(calcTransforms(ary, row_trans, var_cats, ...))
+        }
+
+        ary <- subsetTransformedCube(ary, x)
+    }
+
+    # TODO: calculate column transforms
+
+
+    return(ary)
+}
+
+subsetTransformedCube <- function (ary, cube) {
+    # subset variable categories to only include non-na
+    dims <- dim(ary)
+    keep_all <- lapply(seq_along(dims),
+                       function (i) {
+                           out <- rep(TRUE, dims[i])
+                           names(out) <- dimnames(ary)[[i]]
+                           return(out)
+                       })
+    names(keep_all) <- names(dimensions(cube))[seq_along(keep_all)]
+    keep_these_cube_dims <- evalUseNA(ary, dimensions(cube)[seq_along(keep_all)], cube@useNA)
+
+    # remove the categories determined to be removable above
+    keep_these <- mapply(function(x, y) {
+        not_ys <- y[!y]
+        x[names(not_ys)] <- not_ys
+        return(x)
+    }, keep_all,
+    keep_these_cube_dims,
+    SIMPLIFY=FALSE,
+    USE.NAMES=TRUE)
+
+    # match up those in keep_these that are already in ary
+    ary_dimnames <- dimnames(ary)
+    for (i in seq_along(ary_dimnames)) {
+        to_keep <- names(keep_these[[i]]) %in% ary_dimnames[[i]]
+        keep_these[[i]] <- keep_these[[i]][to_keep]
+    }
+
+    ary <- subsetCubeArray(ary, keep_these)
+
+    return(ary)
+}
+
+#' @rdname Transforms
+#' @export
+setMethod("transforms", "CrunchCube", function (x) {
+    trans <- tryCatch(Transforms(data=index(variables(x))[[1]]$view$transform), error = function(e) NULL)
+    return(trans)
+})
+
+#' @rdname Transforms
+#' @export
+setMethod("transforms<-", c("CrunchCube", "NULL"), function (x, value) {
+    # hacked access to delete the transforms. This should have a more systematic
+    # accessor
+    x@dims@.Data[[1]]$references$view$transform <- value
```

Comment:
`x@dims@.Data` says we have missing methods

`[[1]]` hardcoding isn't necessary here, lapply over

`$references$view$transform` will fail if the "references" has no "view" element, and I don't think there's a guarantee that it will (I see this pattern in a few places)

Reply:
I've added a `dimensions<-` method, so no longer reaching into the object. I've removed the [[1]] hard codings, and added a few extra methods to make the hard coding for rows be all in one place. 

I've also added some checks for `$references$view$transform` view/transform existing. basically, try to select, and use the fact that if any of the named elements don't exist, we get `NULL` (which is used elsewhere with this structure). I've looked at the other uses of that pattern, and they all have similar checks immediately after, so I think we're covered, but if there's a better way to do this, I'm happy to change to that.